docs/ApiGuidelines.md

```diff
@@ -0,0 +1,76 @@
+  API before it is released for the first time.)
+* A public API must not be added to Open Enclave if there is already a non-deprecated API that can just 
+  as easily be used (i.e., with approximately the same number of lines of code).
+* A public API must not be added to Open Enclave if the API would provide generic functionality
+  (i.e., not Open Enclave specific functionality) that could be provided by an external library.
+* All new Public C API names must start with oe_.
+* All public headers must follow Open Enclave [coding conventions](DevelopmentGuide.md).
+
+### Experimental APIs
+
+Experimental APIs are considered to be potential future public APIs, but there is no guarantee 
+of support in future releases, nor is there any guarantee that breaking changes will not occur 
+across releases. 
+
+Experimental API requirements: 
+
+* Experimental APIs must be optional to compile in. That is, there must be some way to build
+  Open Enclave in a way that does not expose any experimental APIs, so that an application
+  can easily verify that it uses only fully-supported public APIs, and so that the Open Enclave
+  code size requirement can be minimized.
+* Experimental APIs must be in a header file that is published to the build directory. They
+  can be in their own header, or in a header file that is shared with public APIs but only
+  if surrounded by an appropriate ifdef.
+* Experimental APIs should be fully documented using doxygen markup in the header file
+* Experimental APIs should ideally be used in some sample code (separate from samples that
+  use only public APIs) or snippet. Ideally, the sample should be one that is compiled as part
+  of the build. For rarely used APIs, the sample might simply be a snippet in the doxygen
+  comments. Having working samples may help others evaluate and determine how ready it is for
+  prime time before the API moves from experimental to public.
+
+### Internal APIs
+
+Internal APIs are considered to be usable within Open Enclave but are not intended to be
+used by applications that use Open Enclave. 
+
+Internal API requirements: 
+
+* Internal APIs must not appear in any header file that is published to the build directory
+* Internal APIs should ideally be documented using doxygen markup in the private header files
+* Internal APIs must not be used in sample code
+
+## Breaking Changes
+
+* Public APIs cannot have source-breaking changes (and, for platforms where shared libraries or built,
+  cannot have binary-breaking changes) across releases, except as covered by deprecation as explained
+  below. API additions can be made at any time.
+* Experimental and internal APIs can have breaking changes across releases.
+
+APIs can be deprecated by marking them as @deprecated. The associated text should explain what
+an application should do instead. Public APIs must be @deprecated in a release before they can be
+removed.
```

Comment:
Does this mean that a public API can't be removed until after its deprecation has been announced first in a given release? Do patch releases count, or only minor/major version releases? Any exceptions to this rule, like for possible security concerns?

Reply:
> Does this mean that a public API can't be removed until after its deprecation has been announced first in a given release?

Yes, that is the proposal.

>  Do patch releases count, or only minor/major version releases? 

Proposal is only minor/major version releases.

> Any exceptions to this rule, like for possible security concerns?

Proposal is no.   Security concerns might motivate making an API have different behavior (e.g., silent success) in some or all conditions, but not removing the API in a way that would cause compiler errors.   Feel free to argue for an alternative proposal, I'm just answering what the proposal is, based on a precedent from another LF project.

Reply:
Breaking public API changes damage the community's trust in a project, so I'd like us to follow a deprecation cycle that's consistent and clear -- and I like that this proposal makes it clear. I suggest that we adopt a strong stance on guaranteeing the stability of public APIs as early as is reasonably possible, but no sooner.

However, if we adhered to [SemVer](https://semver.org/), then a breaking API change -- and that includes the removal of a public stable API -- must always be accompanied by a **major** version change in the API. 

I like the current proposal for now (pre-1.0)

@dthaler could you add a section stating that, when we reach a v1.0, then the project will adopt SemVer?

Reply:
+1 to all of what Aeva said.  How about merging as is and adding such a statement in a separate PR?

Reply:
I opened #2543 to track this

Reply:
I also agree with @AevaOnline and wanted to share that we already SemVer (with the caveat that we're not yet a 1.0 project): 
https://github.com/openenclave/openenclave/blob/master/CHANGELOG.md#changelog
https://github.com/openenclave/openenclave/blob/master/docs/Releasing.md#version-bump

Reply:
I would suggest we document the breaking API change process in more detail
1. Submit a design PR and get it approved. The design PR can minimally state why the old API is problematic and why the new API is needed.
2. Mark API as deprecated in one release only after the design PR is approved.
3. Add new API in the next release.
4. Follow up questions: Can I deprecate in 0.7.2 and replace in 0.8? Can i deprecate in 0.8.2 and replace in 0.8.3?